lib/workers/multi.js

```diff
@@ -30,14 +30,34 @@ class Worker {
   }
 
   /**
+   * Try to get a job via the popper, but if none is available, wait for the
+   * earlier of either 1) the passage of an interval or 2) the completion of
+   * a running job, before trying to pop again.
```

Comment:
This assumes that the completion of a job _may_ enqueue a different job, right? So as long as jobs are churning, someone will keep popping.

Isn't it the case that as soon as work is completed a pop is attempted by the completing worker? This means you're potentially waking up multiple workers when there's only one new job available, right?

Reply:
It doesn't assume that the completion of a job may enqueue a different job. Just that the completion of a job reduces the number of running jobs in that queue, which would make one more job available to be run when using `max-concurrency`.

This provides one pop upon job completion. In a situation where `max-concurrency` is employed, this `pop` should return a job assuming there are jobs queued up (since the completion of the job increments the max-concurrency semaphore by 1, there should be room for one `pop`).

This actually makes the behavior quite similar to that of the `single` worker -- as soon as a job is completed, try to get another. Another way to look at it is that now the the worst case number of `pop`s is `2 * jobs` (one pop to get the job, one pop at completion). In a fully saturated queue (where workload is equal to capacity), this drops to be just `jobs` `pop`s.

Reply:
Ok, I think I got it now. Thanks for the explanation. To restate things, only the worker that _completed_ a job will be woken to immediately pop a new job _from the queue for which a job was just completed_, right?

Reply:
It'll make a call out again to `popper.pop`, but `popper` owns the policy for which queue to pop off of. Currently implemented is just a round-robin popper, which can be configured with multiple queues.

For example, a `multi` worker might be configured to pull from queues `a`, `b`, and `c`. Upon completion of a job (from any of those queues), it will just ask the `popper` for a job, which could conceivably be from any queue.

Reply:
So how will this handle the case if you are hitting `max-concurrency` for *all* queues? It sounds like if you're round-robin across `a`, `b`, and `c`, and all are at `max-concurrency`, you'll block until one of them completes. Let's say `b` completes, but the next queue to be checked in line is `c`. You'll still end up waiting a timeout on `c` unless another job completes. Then you'll wait *another* timeout on `a` unless a job completes and then only when you get back to `b` will you pick up a job even though `b` was ready to pick up a job this whole time.

Reply:
Ah, excellent question. The `popper` round-robins through the queues, but it remains aware of the whether _all_ the queues were empty or if any of them returned a job. Only if _all_ queues had no jobs available will it return a `null` and thus invoke the timeout.

For your example, if a job in `b` completes, we'd call `popper.pop`, which checks `c` and finds nothing. But since its last go-round it did fine a job (in `b`), it will check `a` (and return that job if one is found), `b` (and return a job if one is found), and so on. The delay only gets invoked if no jobs were available on any of the queues.

Another way of framing the change that happened here is that the delay is now whichever comes first: the timeout or a job in the worker completes.